server/src/main/java/org/apache/druid/segment/loading/SegmentLoaderLocalCacheManager.java

```diff
@@ -175,22 +176,33 @@ public File getSegmentFiles(DataSegment segment) throws SegmentLoadingException
    * location may fail because of IO failure, most likely in two cases:<p>
    * 1. druid don't have the write access to this location, most likely the administrator doesn't config it correctly<p>
    * 2. disk failure, druid can't read/write to this disk anymore
+   *
+   * Locations are fetched using {@link StorageLocationSelectorStrategy}.
    */
   private StorageLocation loadSegmentWithRetry(DataSegment segment, String storageDirStr) throws SegmentLoadingException
   {
-    for (StorageLocation loc : locations) {
+    Iterator<StorageLocation> locations = strategy.getLocations(segment, storageDirStr);
+    int numLocationsToTry = this.locations.size();
+
+    while (locations.hasNext() && numLocationsToTry > 0) {
+
+      StorageLocation loc = locations.next();
+      numLocationsToTry--; // This is to avoid the cyclic iterator returned from Round Robin strategy to loop
+      // indefinitely.
+
       File storageDir = loc.reserve(storageDirStr, segment);
-      if (storageDir != null) {
+
+      if (null != storageDir) {
         try {
           loadInLocationWithStartMarker(segment, storageDir);
           return loc;
         }
         catch (SegmentLoadingException e) {
           try {
             log.makeAlert(
-                e,
-                "Failed to load segment in current location [%s], try next location if any",
-                loc.getPath().getAbsolutePath()
+              e,
```

Comment:
Indentation is wrong here. Please don't change indentation.

Reply:
fixed